### PR TITLE
[rtl] Add FPGA Register File

### PR DESCRIPTION
**Pull request by ganoam**
This commit adds a register file designed to be synthesized into FPGA
synchronous-write / asynchronous-read design elements.

For the artya7-100 example, the register file is implemented by 12
RAM32M primitives, conserving approximately 600 Logic LUTs and 1000
flip-flops at the expense of 48 LUTRAMs.

Signed-off-by: ganoam <gnoam@live.com>

### CODE REVIEW

File: dv/uvm/ibex_dv.f [hunk 1]
```diff
@@ -27,7 +27,7 @@
 ${PRJ_DIR}/ibex/rtl/ibex_multdiv_fast.sv
 ${PRJ_DIR}/ibex/rtl/ibex_prefetch_buffer.sv
 ${PRJ_DIR}/ibex/rtl/ibex_fetch_fifo.sv
-${PRJ_DIR}/ibex/rtl/ibex_register_file_ff.sv
+${PRJ_DIR}/ibex/rtl/ibex_register_file_fpga.sv
```
Comment (vogelpi) [inline dv/uvm/ibex_dv.f]: I think we do not want to use the new FPGA Regfile by default. Thus I suggest to undo this change.

File: ibex_core.core [hunk 2]
```diff
@@ -27,7 +27,8 @@ filesets:
       # XXX: Figure out the best way to switch these two implementations
       # dynamically on the target.
 #      - rtl/ibex_register_file_latch.sv # ASIC
-      - rtl/ibex_register_file_ff.sv # FPGA
+#      - rtl/ibex_register_file_ff.sv # FPGA old
+      - rtl/ibex_register_file_fpga.sv # FPGA
```
Comment (vogelpi) [inline ibex_core.core]: I suggest to change the comment for the FF Regfile to "generic FF-based" and to keep using this one by default.

File: rtl/ibex_core.f [hunk 3]
```diff
@@ -12,5 +12,5 @@
 ibex_multdiv_fast.sv
 ibex_prefetch_buffer.sv
 ibex_fetch_fifo.sv
-ibex_register_file_ff.sv
+ibex_register_file_fpga.sv
```
Comment (vogelpi) [inline rtl/ibex_core.f]: Same comment: I think we should keep using the FF-based Regfile by default.

File: rtl/ibex_register_file_fpga.sv [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+  input  logic [DataWidth-1:0] wdata_a_i,
+  input  logic                 we_a_i
+);
+
+  localparam int ADDR_WIDTH = RV32E ? 4 : 5;
+  localparam int NUM_WORDS  = 2**ADDR_WIDTH;
+
+  logic [31:0] mem[NUM_WORDS];
+  logic we; // write enable if writing to any register other than R0
+
```
Comment (vogelpi) [inline rtl/ibex_register_file_fpga.sv]: Remove one newline here.

File: src_files.yml [hunk 5]
```diff
@@ -50,6 +50,6 @@ ibex_regfile_fpga:
   incdirs: [
   ]
   files: [
-    rtl/ibex_register_file_ff.sv,
+    rtl/ibex_register_file_fpga.sv,
```
Comment (vogelpi) [inline src_files.yml]: Same comment as above, keep using the FF-based Regfile by default.

File: rtl/ibex_register_file_fpga.sv [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+  // Write port W1
+  input  logic [          4:0] waddr_a_i,
+  input  logic [DataWidth-1:0] wdata_a_i,
+  input  logic                 we_a_i
+);
+
+  localparam int ADDR_WIDTH = RV32E ? 4 : 5;
+  localparam int NUM_WORDS  = 2**ADDR_WIDTH;
+
+  logic [31:0] mem[NUM_WORDS];
```
Comment (vogelpi) [inline rtl/ibex_register_file_fpga.sv]: Is there a reason why you use an unpacked array here instead of a packed array (`logic [NUM_WORDS-1:0][Datawidth-1:0] mem`)?

File: rtl/ibex_register_file_fpga.sv [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+// Copyright 2018 ETH Zurich and University of Bologna, see also CREDITS.md.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+/**
+* RISC-V register file
+*
+* Register file with 31 or 15x 32 bit wide registers. Register 0 is fixed to 0.
+*
+* This register file is designed to make synthesis tools infer RAM32M primitives for FPGA synthesis.
```
Comment (vogelpi) [inline rtl/ibex_register_file_fpga.sv]: "This register file is designed to make FPGA synthesis tools infer RAM32M primitives."

File: rtl/ibex_register_file_fpga.sv [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+*/
+
+module ibex_register_file #(
+  parameter bit RV32E             = 0 ,
+  parameter int unsigned DataWidth = 32
+) (
+  // Clock and Reset
+  input  logic                 clk_i    ,
+
+  /* verilator lint_off UNUSED */
```
Comment (vogelpi) [inline rtl/ibex_register_file_fpga.sv]: We usually put lint waivers into `lint/verilator_waiver.vlt`. There are already similar rules for `ibex_register_file_ff.sv`.

File: rtl/ibex_register_file_fpga.sv [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+* resulting in a 32x36 dual port RAM. In order to obtain 2 read ports and 1 write port, the 
+* structure is duplicated resulting in 12 RAM32M Instances.
+*/
+
+module ibex_register_file #(
+  parameter bit RV32E             = 0 ,
+  parameter int unsigned DataWidth = 32
+) (
+  // Clock and Reset
+  input  logic                 clk_i    ,
```
Comment (imphil) [inline rtl/ibex_register_file_fpga.sv]: Please remove the space between `clk_i` and the comma (here and below)

File: rtl/ibex_register_file_fpga.sv (hunk 7 above)
Comment (GregAC) [reply rtl/ibex_register_file_fpga.sv]: Presumably this has only been tested with Xilinx? It may work with other vendors too but for now we want to say:

"This register file is designed to make synthesis tools infer RAM primitives for FPGA synthesis. For Xilinx FPGA architectures it will produce RAM32M primitives, other vendors have not yet been tested"

File: rtl/ibex_register_file_fpga.sv [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+
+  localparam int ADDR_WIDTH = RV32E ? 4 : 5;
+  localparam int NUM_WORDS  = 2**ADDR_WIDTH;
+
+  logic [31:0] mem[NUM_WORDS];
+  logic we; // write enable if writing to any register other than R0
+
+
+  always_comb begin : async_read
+    if ( raddr_a_i == '0 ) begin
```
Comment (imphil) [inline rtl/ibex_register_file_fpga.sv]: Please remove the space after `(` and before `)` (here and below)

File: rtl/ibex_register_file_fpga.sv [hunk 11]
```diff
@@ -0,0 +1,76 @@
+// Copyright lowRISC contributors.
+// Copyright 2018 ETH Zurich and University of Bologna, see also CREDITS.md.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+/**
+* RISC-V register file
```
Comment (imphil) [inline rtl/ibex_register_file_fpga.sv]: Please indent the `*` with one space.

```
/**
 * RISC-V register file
...
```

File: rtl/ibex_register_file_fpga.sv (hunk 6 above)
Comment (GregAC) [reply rtl/ibex_register_file_fpga.sv]: I wouldn't be surprised if this style is required to infer the appropriate.

To me it makes more sense too, the packed dimension refers to the individual bits that you can access and manipulate in parallel, whilst the unpacked dimension is for things you access an element at a time.

File: rtl/ibex_register_file_fpga.sv [hunk 12]
```diff
@@ -0,0 +1,76 @@
...
+* Register file with 31 or 15x 32 bit wide registers. Register 0 is fixed to 0.
+*
+* This register file is designed to make synthesis tools infer RAM32M primitives for FPGA synthesis.
+* For the 32 entry register file, 6 RAM32M primitives are used as 32x6 wide simple dual port RAM 
+* resulting in a 32x36 dual port RAM. In order to obtain 2 read ports and 1 write port, the 
+* structure is duplicated resulting in 12 RAM32M Instances.
+*/
+
+module ibex_register_file #(
+  parameter bit RV32E             = 0 ,
```
Comment (imphil) [inline rtl/ibex_register_file_fpga.sv]: either add one more space, or remove the spaces.

File: rtl/ibex_register_file_fpga.sv (hunk 6 above)
Comment (vogelpi) [reply rtl/ibex_register_file_fpga.sv]: I am fine with both and do not have a strong preference here. Just curious. However, `[31:0]` should be replaced by `[DataWidth-1:0]` in any case.

Comment (ganoam) [reply rtl/ibex_register_file_fpga.sv]: No, there is no particular reason. But in the coding example in the synthesis user guide (https://www.xilinx.com/support/documentation/sw_manuals/xilinx2017_1/ug901-vivado-synthesis.pdf, p. 110) it is done like that. I don't know if packed would work as well but it looks more natural to me like this.

I will change the data with declaration.

File: rtl/ibex_register_file_fpga.sv (hunk 7 above)
Comment (ganoam) [reply rtl/ibex_register_file_fpga.sv]: Changed to 
"This register file is designed to make FPGA synthesis tools infer RAM primitives. For Xilinx FPGA architectures, it will produce RAM32M primitives. Other vendors have not yet been tested."

File: dv/uvm/ibex_dv.f (hunk 1 above)
Comment (ganoam) [reply dv/uvm/ibex_dv.f]: Done. Thank you!

File: ibex_core.core (hunk 2 above)
Comment (ganoam) [reply ibex_core.core]: Done. Thank you!

File: rtl/ibex_core.f (hunk 3 above)
Comment (ganoam) [reply rtl/ibex_core.f]: Done. Thank you!

File: rtl/ibex_register_file_fpga.sv (hunk 4 above)
Comment (ganoam) [reply rtl/ibex_register_file_fpga.sv]: Done. Thank you!

File: src_files.yml (hunk 5 above)
Comment (ganoam) [reply src_files.yml]: Done. Thank you!

File: rtl/ibex_register_file_fpga.sv (hunk 9 above)
Comment (ganoam) [reply rtl/ibex_register_file_fpga.sv]: Done. Thank you!

File: rtl/ibex_register_file_fpga.sv (hunk 8 above)
Comment (ganoam) [reply rtl/ibex_register_file_fpga.sv]: Done. Thank you!

File: rtl/ibex_register_file_fpga.sv (hunk 10 above)
Comment (ganoam) [reply rtl/ibex_register_file_fpga.sv]: Done. Thank you!

File: rtl/ibex_register_file_fpga.sv (hunk 6 above)
Comment (vogelpi) [reply rtl/ibex_register_file_fpga.sv]: Perfect, thanks.

File: rtl/ibex_register_file_fpga.sv (hunk 7 above)
Comment (vogelpi) [reply rtl/ibex_register_file_fpga.sv]: Great, thanks for adjusting.

File: rtl/ibex_register_file_fpga.sv (hunk 12 above)
Comment (vogelpi) [reply rtl/ibex_register_file_fpga.sv]: This one is still there.

File: rtl/ibex_register_file_fpga.sv [hunk 13]
```diff
@@ -0,0 +1,70 @@
...
+// Copyright 2018 ETH Zurich and University of Bologna, see also CREDITS.md.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+/**
+ * RISC-V register file
+ *
+ * Register file with 31 or 15x 32 bit wide registers. Register 0 is fixed to 0.
+ *
+ * This register file is designed to make FPGA synthesis tools infer RAM primitives. For Xilinx FPGA
```
Comment (vogelpi) [inline rtl/ibex_register_file_fpga.sv]: Nit: this line has now 101 characters. Please add move "FPGA" to the next line.

File: rtl/ibex_register_file_fpga.sv (hunk 12 above)
Comment (ganoam) [reply rtl/ibex_register_file_fpga.sv]: oups... sorry. Done, thank you.

File: rtl/ibex_register_file_fpga.sv [hunk 14]
```diff
@@ -0,0 +1,70 @@
...
+
+    if (raddr_b_i == '0) begin
+      rdata_b_o = '0;
+    end else begin
+      rdata_b_o = mem[raddr_b_i];
+    end
+  end : async_read
+
+  always_comb begin : we_select
+    if (waddr_a_i == 0) begin
```
Comment (vogelpi) [inline rtl/ibex_register_file_fpga.sv]: Nit: You check vs. `0` here where above, you check vs. `'0`.

Comment (vogelpi) [reply rtl/ibex_register_file_fpga.sv]: Alternatively, you could write this whole block just as:
```
assign we = (waddr_a_i == '0) ? 1'b0 : we_a_i;
```

Comment (ganoam) [reply rtl/ibex_register_file_fpga.sv]: yes, that's a typo. I will correct. 

About the assign statement: The same goes for the async_read blocks. What is the preferred style?

Comment (vogelpi) [reply rtl/ibex_register_file_fpga.sv]: Yeah, you are right. Personally, I prefer the shorter, ternary assignment.

Comment (ganoam) [reply rtl/ibex_register_file_fpga.sv]: Ok. Done :)

File: rtl/ibex_register_file_fpga.sv (hunk 13 above)
Comment (ganoam) [reply rtl/ibex_register_file_fpga.sv]: Done, thanks.
